### PR TITLE
Copy files when they originate out of MEDIA_ROOT

### LINKED ISSUE #5941: Task metrics reporting an incorrect number of unblocked tasks
**Describe the bug**
If a number of unblocked tasks are in the "complete state", like `canceled` or `failed`, somehow they will be showed in the task metrics, leading to some confusing conclusions.

**To Reproduce**
Create some tasks with different exclusive resources, like different `domain`. Make them fail, or cancel them.
It's gonna show a number of unblocked tasks waiting in the queue, which is not true.

**Expected behavior**
It's  expected to only show unblocked tasks that are waiting in the queue to be executed.

### PR DESCRIPTION
**Pull request by bmbouter**
Django would move files in cases where the source was a file:///
repository.

https://pulp.plan.io/issues/5941
closes #5941

https://pulp.plan.io/issues/5870
closes #5870

### CODE REVIEW

File: pulpcore/app/models/storage.py [hunk 1]
```diff
@@ -1,93 +1,38 @@
...
-
-    The non-atomic nature of 3b makes it possible for corrupted Artifacts to be saved if
-    /var/lib/pulp/tmp and /var/lib/pulp/artifact are on separate filesystems. An interrupted save()
-    operation will result in a partial file being placed into /var/lib/pulp/artifact. In such
-    situations, Pulp will not be able to replace the file with the correct version in the
-    future and the user will have to manually remove the partial file.
+    Django's FileSystemStorage with modified _save() and get_available_name behaviors
+
+    The _save() will try to reflink the file first, and then fall back to copy similar to what
+    `cp --reflink=auto` does.
```
Comment (daviddavis) [inline pulpcore/app/models/storage.py]: This needs to be updated.

Comment (bmbouter) [reply pulpcore/app/models/storage.py]: good call. fixed in next push.

File: pulpcore/app/models/storage.py [hunk 2]
```diff
@@ -97,20 +43,56 @@ def save(self, name, content, max_length=None):
...
+                            mode = 'wb' if isinstance(chunk, bytes) else 'wt'
+                            _file = os.fdopen(fd, mode)
+                        _file.write(chunk)
+                finally:
+                    locks.unlock(fd)
+                    if _file is not None:
+                        _file.close()
+                    else:
+                        os.close(fd)
+        except FileExistsError:
```
Comment (dkliban) [inline pulpcore/app/models/storage.py]: Can you refactor this except so that it is more closely located to the code that emits the exception?

Comment (bmbouter) [reply pulpcore/app/models/storage.py]: yes! I pushed that just now.

Comment (bmbouter) [reply pulpcore/app/models/storage.py]: That fails on the open due to the 0o666 open mode, so the exception handler being down here is needed.

**Pull request merged**